staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -129,36 +123,28 @@ func createQueues(n, baseIndex int) []*queue {
 // update handling for when fields are updated is handled here as well -
 // eg: if DesiredNum is increased, SetConfiguration reconciles by
 // adding more queues.
-func (qs *queueSet) SetConfiguration(config fq.QueueSetConfig) error {
+func (qs *queueSet) SetConfiguration(config fq.QueueSetConfig) {
 	qs.lockAndSyncTime()
 	defer qs.lock.Unlock()
-	var dealer *shufflesharding.Dealer
 
-	if config.DesiredNumQueues > 0 {
-		var err error
-		dealer, err = shufflesharding.NewDealer(config.DesiredNumQueues, config.HandSize)
-		if err != nil {
-			return errors.Wrap(err, "shuffle sharding dealer creation failed")
-		}
-		// Adding queues is the only thing that requires immediate action
-		// Removing queues is handled by omitting indexes >DesiredNum from
-		// chooseQueueIndexLocked
-		numQueues := len(qs.queues)
-		if config.DesiredNumQueues > numQueues {
-			qs.queues = append(qs.queues,
-				createQueues(config.DesiredNumQueues-numQueues, len(qs.queues))...)
-		}
-	} else {
+	if config.DesiredNumQueues < 1 {
 		config.QueueLengthLimit = qs.config.QueueLengthLimit
-		config.HandSize = qs.config.HandSize
+		config.Dealer = qs.config.Dealer
```

Comment:
We could continue to error here if `qs.config.Dealer` is nil or set up wrong? I slightly prefer constructing the dealer here, ensuring it can't be wrong. But I'm sure you had a reason for changing this.

Reply:
The problem is visible when you consider the caller's logic, in the filter's configuration controller.  The current logic there calls here after allocating concurrency to the priority level in question.  But if we returned an error here then we would not want to have allocated that concurrency.  So I moved the error detection up before apportioning the concurrency.

Reply:
I understand the problem now. I don't like distributing the dealer construction as a solution, because it means we have to trust people to do it right at every construction site.

Maybe the easiest solution is to take the concurrency field out of the QueueSetConfig struct, and make a separate function here to modify it.

So the caller changes to a two step system:
1. construct and/or change (non-concurrency) settings on all QueueSets; any that fail are removed from consideration for concurrency
2. call SetConcurrency() on all other QueueSets

(Personally I'd probably make an intermediate object, e.g. a CompiledQueueSetConfig which constructs the dealer; once you sucessfully create one of those, you can add concurrency to it and there's no chance of failure when you pass it to QueueSet construction / SetConfiguration. But this is maybe more work.)

Reply:
Yes, I agree with thinking along these lines.